src/Microsoft.OData.Core/Json/JsonReaderExtensions.cs

```diff
@@ -315,6 +317,40 @@ internal static ODataValue ReadAsUntypedOrNullValue(this IJsonReader jsonReader)
             };
         }
 
+        internal static ODataValue ReadODataValue(this IJsonReader jsonReader)
+        {
+            if (jsonReader.NodeType == JsonNodeType.PrimitiveValue)
+            {
+                object primitiveValue = jsonReader.ReadPrimitiveValue();
+
+                return primitiveValue.ToODataValue();
+            }
+            else if (jsonReader.NodeType == JsonNodeType.StartObject)

```

Comment:
>(jsonReader.NodeType == JsonNodeType.StartObject) [](start = 20, length = 49)

how about the StartArray NodeType

Reply:
So, if the payload contains an array, we will attempt to read using ReadAsUntypedOrNullValue -- is that the expected/desired behavior?

---
In reply to: [320876548](https://github.com/OData/odata.net/pull/1523#discussion_r320876548) [](ancestors = 320876548)